src/render/program/line_program.js

```diff
@@ -181,13 +139,22 @@ function calculateMatrix(painter, tile, layer, matrix) {
     );
 }
 
+const lineDefinesValues = (layer: LineStyleLayer): LineDefinesType[] => {
+    const values = [];
+    if (hasDash(layer)) values.push('RENDER_LINE_DASH');
+    if (layer.paint.get('line-gradient')) values.push('RENDER_LINE_GRADIENT');
+    return values;
```

Comment:
Not a blocker for this PR and could be deferred, but wondering if these new shader variants are still compiled as part of the shader pre-compilation step (introduced by commit https://github.com/mapbox/mapbox-gl-js/pull/10160/commits/b93d70979ef832e5fc0ad401253e3b50fefff4d6), if it's not we should add a follow up ticket to look at that.

Reply:
I've checked it now and I think it does. At least in `painter.js` I can see the program in the `cache` and the keying also depends on these define values. (I've actually written this based on the `circleDefinesValues` which are used in a similar way.)

Reply:
Thanks for checking that one! That makes sense.